File: Intatis-Windows/src/Intatis.Core/Providers/ConfigImport.cs
using System.Text.Json.Nodes;

namespace Intatis.Core.Providers;

public enum SecretSource
{
    Environment,
    File,
    AuthFile,
    ProviderConfig,
    Literal,
    None,
}

/// <summary>
/// A credential reference, never a credential value. Secrets resolve lazily at
/// provider-construction time and never enter the event log or projections.
/// </summary>
public sealed record SecretRef
{
    public SecretSource Source { get; init; } = SecretSource.None;
    public string Value { get; init; } = "";

    public string Describe() => Source switch
    {
        SecretSource.Environment => $"env {Value}",
        SecretSource.File => $"secret file {Value}",
        SecretSource.AuthFile => "auth file",
        SecretSource.ProviderConfig => "provider config",
        SecretSource.Literal => "configured key",
        _ => "not configured",
    };
}

public sealed record ModelEntry
{
    public string Id { get; init; } = "";
    public string DisplayName { get; init; } = "";
    public bool Hidden { get; init; } // role-routed models stay out of the inference menus
}

public sealed record ProviderEntry
{
    public string Id { get; init; } = "";
    public string DisplayName { get; init; } = "";
    public string BaseUrl { get; init; } = "";
    public string? ChatEndpoint { get; init; }
    public SecretRef ApiKeyRef { get; init; } = new();
    public List<ModelEntry> Models { get; init; } = [];
    public string Npm { get; init; } = "@ai-sdk/openai-compatible";
}

/// <summary>&lt;provider-id&gt;/&lt;model-id&gt; role binding.</summary>
public sealed record ModelRef
{
    public string ProviderId { get; init; } = "";
    public string ModelId { get; init; } = "";
}

public sealed record ImportedConfig
{
    public List<ProviderEntry> Providers { get; init; } = [];
    public ModelRef? Chat { get; init; }
    public ModelRef? Reviewer { get; init; }
    public ModelRef? Image { get; init; }
    public ModelRef? Transcription { get; init; }
    public ModelRef? Embedding { get; init; }
    public ModelRef? Reranker { get; init; }
    public string SourcePath { get; init; } = "";
    public List<string> Warnings { get; init; } = [];
    public bool ReviewerFailedClosed { get; init; }

    public ProviderEntry? Provider(string id)
        => Providers.FirstOrDefault(p => string.Equals(p.Id, id, StringComparison.OrdinalIgnoreCase));

    public List<ModelEntry> InferenceModels()
    {
        var result = new List<ModelEntry>();
        foreach (var provider in Providers)
        {
            foreach (var model in provider.Models)
            {
                if (model.Hidden) continue;
                result.Add(model);
            }
        }
        return result;
    }
}

/// <summary>
/// Parses the Intatis JSON/JSONC configuration: the provider map plus the top-level
/// role routes (model, permission_reviewer_model, image_model, transcription_model,
/// embedding_model, reranker_model). The permission reviewer binding is fail-closed:
/// if the field is present but unresolvable, reviewing stays unavailable instead of
/// silently falling back to the chat model.
/// </summary>
public static class ConfigImport
{
    public const int MaximumByteCount = 1_048_576;

    private static readonly Dictionary<string, string> DefaultBaseUrls = new(StringComparer.OrdinalIgnoreCase)
    {
        ["openai"] = "https://api.openai.com/v1",
        ["openrouter"] = "https://openrouter.ai/api/v1",
        ["deepseek"] = "https://api.deepseek.com/v1",
        ["ollama"] = "http://localhost:11434/v1",
        ["lmstudio"] = "http://localhost:1234/v1",
        ["groq"] = "https://api.groq.com/openai/v1",
        ["xai"] = "https://api.x.ai/v1",
        ["together"] = "https://api.together.xyz/v1",
        ["fireworks"] = "https://api.fireworks.ai/inference/v1",
        ["cerebras"] = "https://api.cerebras.ai/v1",
        ["moonshot"] = "https://api.moonshot.cn/v1",
    };

    public static ImportedConfig Parse(string source, string sourcePath, IReadOnlyDictionary<string, string>? environment = null)
    {
        if (source.Length > MaximumByteCount)
            throw new InvalidOperationException("configuration exceeds 1 MiB");
        var root = Jsonx.ParseObject(Jsonx.StripJsonc(source));
        environment ??= new Dictionary<string, string>();

        var warnings = new List<string>();

        var enabled = ReadStringList(root, "enabled_providers", "enabledProviders");
        var disabled = ReadStringList(root, "disabled_providers", "disabledProviders")
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

        var providers = new List<ProviderEntry>();
        if (root["provider"] is JsonObject providerMap)
        {
            foreach (var (providerId, providerNode) in providerMap)
            {
                if (providerNode is null) continue;
                if (disabled.Contains(providerId)) continue;
                if (enabled.Count > 0 &&
                    !enabled.Contains(providerId, StringComparer.OrdinalIgnoreCase)) continue;

                var entry = ParseProvider(providerId, providerNode, environment, warnings);
                if (entry is null)
                {
                    warnings.Add($"provider '{providerId}' skipped: no resolvable base URL");
                    continue;
                }
                providers.Add(entry);
            }
        }
        else if (root["providers"] is JsonArray legacyArray)
        {
            foreach (var node in legacyArray)
            {
                if (node is not JsonObject providerNode) continue;
                var id = (string?)providerNode["id"] ?? "";
                if (id.Length == 0) continue;
                var entry = ParseProvider(id, providerNode, environment, warnings);
                if (entry is null)
                {
                    warnings.Add($"provider '{id}' skipped: no resolvable base URL");
                    continue;
                }
                providers.Add(entry);
            }
        }

        var chat = ParseModelRef(root, "model", environment)
                   ?? ParseModelRef(root, "small_model", environment)
                   ?? ParseModelRef(root, "smallModel", environment);

        ModelRef? reviewer;
        var reviewerFailedClosed = false;
        if (root["permission_reviewer_model"] is { } reviewerNode)
        {
            reviewer = ParseModelRef(root, "permission_reviewer_model", environment);
            if (reviewer is null || Resolve(reviewer, providers) is null)
            {
                reviewer = null;
                reviewerFailedClosed = true;
                warnings.Add("permission_reviewer_model present but unresolvable; automatic review is disabled (fail closed)");
            }
        }
        else
        {
            // Compatibility: a missing field inherits the same document's top-level model.
            reviewer = chat is not null && Resolve(chat, providers) is not null ? chat : null;
        }

        var image = OptionalRole(root, "image_model", environment, providers, warnings, hiddenRole: true);
        var transcription = OptionalRole(root, "transcription_model", environment, providers, warnings, hiddenRole: true);
        var embedding = OptionalRole(root, "embedding_model", environment, providers, warnings, hiddenRole: true);
        var reranker = OptionalRole(root, "reranker_model", environment, providers, warnings, hiddenRole: true);

        MarkRoleRoutedModelsHidden(providers, chat, reviewer,
            new[] { image, transcription, embedding, reranker });

        if (chat is null && providers.Count > 0)
            warnings.Add("no top-level 'model' field; the active model must be picked in the UI");

        return new ImportedConfig
        {
            Providers = providers,
            Chat = chat,
            Reviewer = reviewer,
            Image = image,
            Transcription = transcription,
            Embedding = embedding,
            Reranker = reranker,
            SourcePath = sourcePath,
            Warnings = warnings,
            ReviewerFailedClosed = reviewerFailedClosed,
        };
    }

    /// <summary>
    /// Models bound only to extension roles (image/transcription/embedding/reranker)
    /// stay out of the Chat/Code/Cowork inference menus; models referenced by the
    /// chat or reviewer routes remain selectable.
    /// </summary>
    private static void MarkRoleRoutedModelsHidden(
        List<ProviderEntry> providers,
        ModelRef? chat,
        ModelRef? reviewer,
        ModelRef?[] extensionRoles)
    {
        var extensionKeys = extensionRoles
            .Where(r => r is not null)
            .Select(r => (r!.ProviderId.ToLowerInvariant(), r.ModelId))
            .ToHashSet();

        var menuKeys = new HashSet<(string, string)>(StringComparer.Ordinal);
        void Keep(ModelRef? r)
        {
            if (r is null) return;
            menuKeys.Add((r.ProviderId.ToLowerInvariant(), r.ModelId));
        }
        Keep(chat);
        Keep(reviewer);

        foreach (var provider in providers)
        {
            for (var i = 0; i < provider.Models.Count; i++)
            {
                var model = provider.Models[i];
                var key = (provider.Id.ToLowerInvariant(), model.Id);
                if (extensionKeys.Contains(key) && !menuKeys.Contains(key))
                    provider.Models[i] = model with { Hidden = true };
            }
        }
    }

    private static ProviderEntry? ParseProvider(
        string providerId,
        JsonNode node,
        IReadOnlyDictionary<string, string> environment,
        List<string> warnings)
    {
        var o = node as JsonObject;
        var options = o?["options"] as JsonObject;

        var baseUrl = (string?)options?["baseURL"] ?? (string?)options?["baseUrl"]
            ?? (string?)o?["baseURL"] ?? (string?)o?["baseUrl"]
            ?? DefaultBaseUrls.GetValueOrDefault(providerId) ?? "";
        if (baseUrl.Length == 0) return null;

        var chatEndpoint = (string?)options?["chatEndpoint"] ?? (string?)o?["chatEndpoint"];
        var displayName = (string?)o?["displayName"] ?? (string?)o?["name"] ?? providerId;
        var npm = (string?)o?["npm"] ?? "@ai-sdk/openai-compatible";

        var apiKeyRef = ParseCredential(providerId, options, o, environment);

        var models = new List<ModelEntry>();
        if (o?["models"] is JsonObject modelMap)
        {
            foreach (var (modelId, modelNode) in modelMap)
            {
                if (modelNode is null) continue;
                if (modelNode is JsonValue value && value.TryGetValue<string>(out var modelName))
                {
                    models.Add(new ModelEntry { Id = modelId, DisplayName = modelName });
                    continue;
                }
                if (modelNode is JsonObject modelObj)
                {
                    models.Add(new ModelEntry
                    {
                        Id = (string?)modelObj["id"] ?? modelId,
                        DisplayName = (string?)modelObj["displayName"]
                                      ?? (string?)modelObj["name"] ?? modelId,
                    });
                }
            }
        }
        else if (o?["models"] is JsonArray modelArray)
        {
            foreach (var modelNode in modelArray)
            {
                var id = (string?)modelNode?["id"] ?? (string?)modelNode;
                if (id is null) continue;
                models.Add(new ModelEntry
                {
                    Id = id,
                    DisplayName = (string?)modelNode?["displayName"] ?? (string?)modelNode?["name"] ?? id,
                });
            }
        }

        return new ProviderEntry
        {
            Id = providerId,
            DisplayName = displayName,
            BaseUrl = baseUrl,
            ChatEndpoint = chatEndpoint,
            ApiKeyRef = apiKeyRef,
            Models = models,
            Npm = npm,
        };
    }

    private static SecretRef ParseCredential(
        string providerId,
        JsonObject? options,
        JsonObject? providerObj,
        IReadOnlyDictionary<string, string> environment)
    {
        if (providerObj?["apiKeySource"] is JsonObject apiKeySource)
        {
            var type = ((string?)apiKeySource["type"] ?? "").ToLowerInvariant();
            var value = (string?)apiKeySource["value"] ?? "";
            return type switch
            {
                "env" => new SecretRef { Source = SecretSource.Environment, Value = value },
                "file" => new SecretRef { Source = SecretSource.File, Value = value },
                "authfile" => new SecretRef { Source = SecretSource.AuthFile, Value = providerId },
                "providerconfig" => new SecretRef { Source = SecretSource.ProviderConfig, Value = providerId },
                _ => new SecretRef { Source = SecretSource.AuthFile, Value = providerId },
            };
        }

        var apiKey = (string?)options?["apiKey"] ?? (string?)providerObj?["apiKey"];
        if (apiKey is { Length: > 0 })
        {
            var variable = ParseConfigVariable(apiKey);
            if (variable is { Kind: "env" })
                return new SecretRef { Source = SecretSource.Environment, Value = variable.Value };
            if (variable is { Kind: "file" })
                return new SecretRef { Source = SecretSource.File, Value = variable.Value };
            return new SecretRef { Source = SecretSource.Literal, Value = apiKey };
        }

        var apiKeyEnv = (string?)providerObj?["apiKeyEnv"]
                        ?? (providerObj?["env"] as JsonArray)?.OfType<JsonValue>()
                            .Select(v => (string?)v).FirstOrDefault(v => v is not null);
        if (apiKeyEnv is { Length: > 0 })
            return new SecretRef { Source = SecretSource.Environment, Value = apiKeyEnv };

        var apiKeyFile = (string?)providerObj?["apiKeyFile"];
        if (apiKeyFile is { Length: > 0 })
            return new SecretRef { Source = SecretSource.File, Value = apiKeyFile };

        return new SecretRef { Source = SecretSource.AuthFile, Value = providerId };
    }

    private static ModelRef? ParseModelRef(
        JsonObject root,
        string field,
        IReadOnlyDictionary<string, string> environment)
    {
        var raw = root[field] as string;
        if (raw is null or { Length: 0 }) return null;

        var variable = ParseConfigVariable(raw);
        if (variable is { Kind: "env" } && environment.TryGetValue(variable.Value, out var resolved))
            raw = resolved;
        else if (variable is { Kind: "env" })
            return null;

        var slash = raw.IndexOf('/');
        if (slash <= 0 || slash >= raw.Length - 1) return null;
        return new ModelRef { ProviderId = raw[..slash], ModelId = raw[(slash + 1)..] };
    }

    private static ModelRef? OptionalRole(
        JsonObject root,
        string field,
        IReadOnlyDictionary<string, string> environment,
        List<ProviderEntry> providers,
        List<string> warnings,
        bool hiddenRole)
    {
        var reference = ParseModelRef(root, field, environment);
        if (reference is null)
        {
            if (root[field] is string present)
                warnings.Add($"{field} present but not in '<provider>/<model>' form; route disabled");
            return null;
        }
        var provider = providers.FirstOrDefault(p =>
            string.Equals(p.Id, reference.ProviderId, StringComparison.OrdinalIgnoreCase));
        if (provider is null)
        {
            warnings.Add($"{field} references unknown provider '{reference.ProviderId}'; route disabled");
            return null;
        }
        return reference;
    }

    internal static ProviderEntry? Resolve(ModelRef reference, List<ProviderEntry> providers)
        => providers.FirstOrDefault(p =>
            string.Equals(p.Id, reference.ProviderId, StringComparison.OrdinalIgnoreCase)
            && p.Models.Any(m => string.Equals(m.Id, reference.ModelId, StringComparison.Ordinal)));

    /// <summary>Whole trimmed value must be {kind:value}; kinds are case-insensitive.</summary>
    internal static (string Kind, string Value)? ParseConfigVariable(string value)
    {
        var trimmed = value.Trim();
        if (trimmed.Length < 3 || trimmed[0] != '{' || trimmed[^1] != '}') return null;
        var inner = trimmed[1..^1];
        var colon = inner.IndexOf(':');
        if (colon <= 0 || colon >= inner.Length - 1) return null;
        var kind = inner[..colon].Trim().ToLowerInvariant();
        var body = inner[(colon + 1)..].Trim();
        if (kind is not ("env" or "file")) return null;
        return (kind, body);
    }

    private static List<string> ReadStringList(JsonObject root, params string[] fields)
    {
        foreach (var field in fields)
        {
            if (root[field] is JsonArray array)
            {
                return array.OfType<JsonValue>()
                    .Select(v => (string?)v)
                    .Where(v => v is not null)
                    .Select(v => v!.Trim().ToLowerInvariant())
                    .ToList();
            }
        }
        return [];
    }
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/providers/ConfigImport.kt
package com.intatis.shared.providers

import com.intatis.shared.protocol.Jsonx
import com.intatis.shared.protocol.Jsonx.str
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

enum class SecretSource { ENVIRONMENT, FILE, AUTH_FILE, PROVIDER_CONFIG, LITERAL, NONE }

/**
 * A credential reference, never a credential value. Secrets resolve lazily at
 * provider-construction time and never enter the event log or projections.
 */
data class SecretRef(val source: SecretSource = SecretSource.NONE, val value: String = "") {
    fun describe(): String = when (source) {
        SecretSource.ENVIRONMENT -> "env $value"
        SecretSource.FILE -> "secret file $value"
        SecretSource.AUTH_FILE -> "auth file"
        SecretSource.PROVIDER_CONFIG -> "provider config"
        SecretSource.LITERAL -> "configured key"
        SecretSource.NONE -> "not configured"
    }
}

data class ModelEntry(val id: String, val displayName: String = "", val hidden: Boolean = false)

data class ProviderEntry(
    val id: String,
    val displayName: String = id,
    val baseUrl: String = "",
    val chatEndpoint: String? = null,
    val apiKeyRef: SecretRef = SecretRef(),
    val models: MutableList<ModelEntry> = mutableListOf(),
    val npm: String = "@ai-sdk/openai-compatible",
)

/** <provider-id>/<model-id> role binding. */
data class ModelRef(val providerId: String, val modelId: String) {
    val displayLabel: String get() = "$providerId/$modelId"
}

class ImportedConfig(
    var providers: MutableList<ProviderEntry> = mutableListOf(),
    var chat: ModelRef? = null,
    var reviewer: ModelRef? = null,
    var image: ModelRef? = null,
    var transcription: ModelRef? = null,
    var embedding: ModelRef? = null,
    var reranker: ModelRef? = null,
    var sourcePath: String = "",
    var warnings: MutableList<String> = mutableListOf(),
    var reviewerFailedClosed: Boolean = false,
) {
    fun provider(id: String): ProviderEntry? =
        providers.firstOrNull { it.id.equals(id, ignoreCase = true) }

    fun inferenceModels(): List<ModelEntry> = providers.flatMap { p -> p.models.filter { !it.hidden } }
}

/**
 * Parses the Intatis JSON/JSONC configuration: the provider map plus the top-level
 * role routes (model, permission_reviewer_model, image_model, transcription_model,
 * embedding_model, reranker_model). The reviewer binding is fail-closed when the
 * field is present but unresolvable.
 */
object ConfigImport {
    const val MAXIMUM_BYTE_COUNT = 1_048_576

    private val defaultBaseUrls = mapOf(
        "openai" to "https://api.openai.com/v1",
        "openrouter" to "https://openrouter.ai/api/v1",
        "deepseek" to "https://api.deepseek.com/v1",
        "ollama" to "http://localhost:11434/v1",
        "lmstudio" to "http://localhost:1234/v1",
        "groq" to "https://api.groq.com/openai/v1",
        "xai" to "https://api.x.ai/v1",
        "together" to "https://api.together.xyz/v1",
        "fireworks" to "https://api.fireworks.ai/inference/v1",
        "cerebras" to "https://api.cerebras.ai/v1",
        "moonshot" to "https://api.moonshot.cn/v1",
    )

    fun parse(source: String, sourcePath: String, environment: Map<String, String> = emptyMap()): ImportedConfig {
        require(source.length <= MAXIMUM_BYTE_COUNT) { "configuration exceeds 1 MiB" }
        val root = Jsonx.parseObject(Jsonx.stripJsonc(source))
        val warnings = mutableListOf<String>()

        val enabled = readStringList(root, "enabled_providers", "enabledProviders")
            .map { it.lowercase() }.toSet()
        val disabled = readStringList(root, "disabled_providers", "disabledProviders")
            .map { it.lowercase() }.toSet()

        val providers = mutableListOf<ProviderEntry>()
        val providerMap = root["provider"] as? JsonObject
        if (providerMap != null) {
            for ((providerId, providerNode) in providerMap) {
                if (providerNode == null) continue
                if (providerId.lowercase() in disabled) continue
                if (enabled.isNotEmpty() && providerId.lowercase() !in enabled) continue
                val entry = parseProvider(providerId, providerNode, environment, warnings) ?: run {
                    warnings.add("provider '$providerId' skipped: no resolvable base URL")
                    continue
                }
                providers.add(entry)
            }
        } else if (root["providers"] is JsonArray) {
            for (node in root["providers"] as JsonArray) {
                val obj = node as? JsonObject ?: continue
                val id = obj.str("id") ?: ""
                if (id.isEmpty()) continue
                val entry = parseProvider(id, obj, environment, warnings) ?: run {
                    warnings.add("provider '$id' skipped: no resolvable base URL")
                    continue
                }
                providers.add(entry)
            }
        }

        val chat = parseModelRef(root, "model", environment)
            ?: parseModelRef(root, "small_model", environment)
            ?: parseModelRef(root, "smallModel", environment)

        var reviewer: ModelRef? = null
        var reviewerFailedClosed = false
        if (root["permission_reviewer_model"] != null) {
            reviewer = parseModelRef(root, "permission_reviewer_model", environment)
            if (reviewer == null || resolve(reviewer, providers) == null) {
                reviewer = null
                reviewerFailedClosed = true
                warnings.add("permission_reviewer_model present but unresolvable; automatic review is disabled (fail closed)")
            }
        } else {
            // Compatibility: a missing field inherits the same document's top-level model.
            reviewer = if (chat != null && resolve(chat, providers) != null) chat else null
        }

        val image = optionalRole(root, "image_model", environment, providers, warnings)
        val transcription = optionalRole(root, "transcription_model", environment, providers, warnings)
        val embedding = optionalRole(root, "embedding_model", environment, providers, warnings)
        val reranker = optionalRole(root, "reranker_model", environment, providers, warnings)

        markRoleRoutedModelsHidden(providers, chat, reviewer, listOf(image, transcription, embedding, reranker))

        if (chat == null && providers.isNotEmpty()) {
            warnings.add("no top-level 'model' field; the active model must be picked in the UI")
        }

        return ImportedConfig(
            providers = providers,
            chat = chat,
            reviewer = reviewer,
            image = image,
            transcription = transcription,
            embedding = embedding,
            reranker = reranker,
            sourcePath = sourcePath,
            warnings = warnings,
            reviewerFailedClosed = reviewerFailedClosed,
        )
    }

    private fun parseProvider(
        providerId: String,
        node: JsonElement,
        environment: Map<String, String>,
        warnings: MutableList<String>,
    ): ProviderEntry? {
        val o = node as? JsonObject ?: return null
        val options = o["options"] as? JsonObject

        val baseUrl = options?.str("baseURL")
            ?: options?.str("baseUrl")
            ?: o.str("baseURL")
            ?: o.str("baseUrl")
            ?: defaultBaseUrls[providerId.lowercase()]
            ?: ""
        if (baseUrl.isEmpty()) return null

        val entry = ProviderEntry(
            id = providerId,
            displayName = o.str("displayName") ?: o.str("name") ?: providerId,
            baseUrl = baseUrl,
            chatEndpoint = options?.str("chatEndpoint") ?: o.str("chatEndpoint"),
            npm = o.str("npm") ?: "@ai-sdk/openai-compatible",
            apiKeyRef = parseCredential(providerId, options, o, environment),
        )

        val modelMap = o["models"] as? JsonObject
        if (modelMap != null) {
            for ((modelId, modelNode) in modelMap) {
                if (modelNode is JsonPrimitive) {
                    entry.models.add(ModelEntry(id = modelId, displayName = modelNode.content))
                    continue
                }
                val modelObj = modelNode as? JsonObject ?: continue
                entry.models.add(ModelEntry(
                    id = modelObj.str("id") ?: modelId,
                    displayName = modelObj.str("displayName") ?: modelObj.str("name") ?: modelId,
                ))
            }
        } else {
            val modelArray = o["models"] as? JsonArray
            modelArray?.forEach { modelNode ->
                val id = (modelNode as? JsonObject)?.str("id")
                    ?: (modelNode as? JsonPrimitive)?.content
                    ?: return@forEach
                val obj = modelNode as? JsonObject
                entry.models.add(ModelEntry(
                    id = id,
                    displayName = obj?.str("displayName") ?: obj?.str("name") ?: id,
                ))
            }
        }
        return entry
    }

    private fun parseCredential(
        providerId: String,
        options: JsonObject?,
        providerObj: JsonObject?,
        environment: Map<String, String>,
    ): SecretRef {
        val apiKeySource = providerObj?.get("apiKeySource") as? JsonObject
        if (apiKeySource != null) {
            return when (val type = (apiKeySource.str("type") ?: "").lowercase()) {
                "env" -> SecretRef(SecretSource.ENVIRONMENT, apiKeySource.str("value") ?: "")
                "file" -> SecretRef(SecretSource.FILE, apiKeySource.str("value") ?: "")
                "providerconfig" -> SecretRef(SecretSource.PROVIDER_CONFIG, providerId)
                else -> SecretRef(SecretSource.AUTH_FILE, providerId)
            }
        }

        val apiKey = options?.str("apiKey") ?: providerObj?.str("apiKey")
        if (!apiKey.isNullOrEmpty()) {
            val variable = parseConfigVariable(apiKey)
            return when {
                variable?.first == "env" -> SecretRef(SecretSource.ENVIRONMENT, variable.second)
                variable?.first == "file" -> SecretRef(SecretSource.FILE, variable.second)
                else -> SecretRef(SecretSource.LITERAL, apiKey)
            }
        }

        val envArray = providerObj?.get("env") as? JsonArray
        val apiKeyEnv = providerObj?.str("apiKeyEnv")
            ?: envArray?.firstOrNull()?.let { (it as? JsonPrimitive)?.content }
        if (!apiKeyEnv.isNullOrEmpty()) return SecretRef(SecretSource.ENVIRONMENT, apiKeyEnv)

        val apiKeyFile = providerObj?.str("apiKeyFile")
        if (!apiKeyFile.isNullOrEmpty()) return SecretRef(SecretSource.FILE, apiKeyFile)

        return SecretRef(SecretSource.AUTH_FILE, providerId)
    }

    private fun parseModelRef(root: JsonObject, field: String, environment: Map<String, String>): ModelRef? {
        var raw = root.str(field) ?: return null
        val variable = parseConfigVariable(raw)
        if (variable?.first == "env") {
            raw = environment[variable.second] ?: return null
        }
        val slash = raw.indexOf('/')
        if (slash <= 0 || slash >= raw.length - 1) return null
        return ModelRef(providerId = raw.substring(0, slash), modelId = raw.substring(slash + 1))
    }

    private fun optionalRole(
        root: JsonObject,
        field: String,
        environment: Map<String, String>,
        providers: List<ProviderEntry>,
        warnings: MutableList<String>,
    ): ModelRef? {
        val reference = parseModelRef(root, field, environment)
        if (reference == null) {
            if (root[field] != null) warnings.add("$field present but not in '<provider>/<model>' form; route disabled")
            return null
        }
        val provider = providers.firstOrNull { it.id.equals(reference.providerId, ignoreCase = true) }
        if (provider == null) {
            warnings.add("$field references unknown provider '${reference.providerId}'; route disabled")
            return null
        }
        return reference
    }

    internal fun resolve(reference: ModelRef, providers: List<ProviderEntry>): ProviderEntry? =
        providers.firstOrNull {
            it.id.equals(reference.providerId, ignoreCase = true) &&
                it.models.any { m -> m.id == reference.modelId }
        }

    /** Whole trimmed value must be {kind:value}; kinds are case-insensitive. */
    internal fun parseConfigVariable(value: String): Pair<String, String>? {
        val trimmed = value.trim()
        if (trimmed.length < 3 || trimmed[0] != '{' || trimmed[trimmed.length - 1] != '}') return null
        val inner = trimmed.substring(1, trimmed.length - 1)
        val colon = inner.indexOf(':')
        if (colon <= 0 || colon >= inner.length - 1) return null
        val kind = inner.substring(0, colon).trim().lowercase()
        val body = inner.substring(colon + 1).trim()
        if (kind !in setOf("env", "file")) return null
        return kind to body
    }

    /** Models bound only to extension roles stay out of the chat inference menus. */
    private fun markRoleRoutedModelsHidden(
        providers: MutableList<ProviderEntry>,
        chat: ModelRef?,
        reviewer: ModelRef?,
        extensionRoles: List<ModelRef?>,
    ) {
        val extensionKeys = extensionRoles.filterNotNull()
            .map { it.providerId.lowercase() to it.modelId }.toHashSet()
        val menuKeys = buildSet {
            chat?.let { add(it.providerId.lowercase() to it.modelId) }
            reviewer?.let { add(it.providerId.lowercase() to it.modelId) }
        }
        for (provider in providers) {
            for (i in provider.models.indices) {
                val model = provider.models[i]
                val key = provider.id.lowercase() to model.id
                if (key in extensionKeys && key !in menuKeys) {
                    provider.models[i] = model.copy(hidden = true)
                }
            }
        }
    }

    private fun readStringList(root: JsonObject, vararg fields: String): List<String> {
        for (field in fields) {
            val array = root[field] as? JsonArray ?: continue
            return array.mapNotNull { (it as? JsonPrimitive)?.content?.trim()?.lowercase() }
        }
        return emptyList()
    }
}
